frappe/public/js/frappe/form/controls/time.js

```diff
@@ -1,42 +1,87 @@
+
+		this.time_format = frappe.defaultTimeFormat;
+		this.datepicker_options = {
 			language: "en",
 			timepicker: true,
 			onlyTimepicker: true,
-			timeFormat: "hh:ii:ss",
+			timeFormat: time_format.toLowerCase().replace("mm", "ii"),
 			startDate: frappe.datetime.now_time(true),
-			onSelect: function() {
+			onSelect: () => {
 				// ignore micro seconds
-				if (moment(me.get_value(), 'hh:mm:ss').format('HH:mm:ss') != moment(me.value, 'hh:mm:ss').format('HH:mm:ss')) {
-					me.$input.trigger('change');
+				if (moment(this.get_value(), time_format).format('HH:mm:ss') != moment(this.value, time_format).format('HH:mm:ss')) {
+					this.$input.trigger('change');
 				}				
 			},
-			onShow: function() {
+			onShow: () => {
 				$('.datepicker--button:visible').text(__('Now'));
+
+				this.update_datepicker_position();
 			},
 			keyboardNav: false,
 			todayButton: true
-		});
+		};
+	},
+	set_input: function(value) {
+		this._super(value);
+		if (value
+			&& ((this.last_value && this.last_value !== this.value)
+				|| (!this.datepicker.selectedDates.length))) {
+
+			var date_obj = frappe.datetime.moment_to_date_obj(moment(value, frappe.sys_defaults['time_format']));
+			this.datepicker.selectDate(date_obj);
+		}
+	},
+	set_datepicker: function() {
+		this.$input.datepicker(this.datepicker_options);
 		this.datepicker = this.$input.data('datepicker');
+
 		this.datepicker.$datepicker
 			.find('[data-action="today"]')
 			.click(() => {
 				this.datepicker.selectDate(frappe.datetime.now_time(true));
 			});
-		this.refresh();
+		this.datepicker.opts['offset'] = 0;
+		if (this.datepicker.opts.timeFormat.indexOf('s') == -1) {
+			// No seconds in time format
+			this.noSeconds = true;
+			const $tp = this.datepicker.timepicker;
+			$tp.$seconds.parent().css('display', 'none');
+			$tp.$secondsText.css('display', 'none');
+			$tp.$secondsText.prev().css('display', 'none');
+		} else {
+			this.noSeconds = false;
+		}
 	},
-	set_input: function(value) {
-		this._super(value);
-		if(value
-			&& ((this.last_value && this.last_value !== this.value)
-				|| (!this.datepicker.selectedDates.length))) {
+	update_datepicker_position: function() {
+		if (!this.frm) return;
+		// show datepicker above or below the input
+		// based on scroll position
+		const scroll_limit = $(window).scrollTop() + $(window).height();
+		const picker_bottom = this.$input.offset().top + this.$input.outerHeight() + 12 + this.datepicker.$datepicker.outerHeight();
 
-			var date_obj = frappe.datetime.moment_to_date_obj(moment(value, 'HH:mm:ss'));
-			this.datepicker.selectDate(date_obj);
+		var position = 'bottom left';
+		//if (picker_top > scroll_limit + window_scroll) {
+		if (picker_bottom >= scroll_limit) {
+			position = 'top left';
+			// bodge around the picker being 30 pixels shorter than
+			// it should be due to hiding seconds
+			if (this.noSeconds) this.datepicker.opts['offset'] = -28;
+		} else {
+			if (this.noSeconds) this.datepicker.opts['offset'] = 12;
```

Comment:
I think `time.js` can be extended from `date.js` so we can share the `update_datepicker_position` code.

Reply:
New plan is to only put update_datepicker_position in date.js, and extend from there to time.js and datetime.js, as suggested.